plumbing/format/commitgraph/v2/chain.go

```diff
@@ -0,0 +1,100 @@
+package v2
+
+import (
+	"bufio"
+	"io"
+	"path"
+
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+type closers []io.Closer
+
+func (c closers) Close() error {
+	var err error
+	for _, closer := range c {
+		currentErr := closer.Close()
+		if currentErr != nil {
+			err = currentErr
+		}
+	}
+	return err
+}
+
+// OpenChainFile reads a commit chain file and returns a slice of the hashes within it
+//
+// Commit-Graph chains are described at https://git-scm.com/docs/commit-graph
+// and are new line separated list of graph file hashes, oldest to newest.
+//
+// This function simply reads the file and returns the hashes as a slice.
+func OpenChainFile(r io.Reader) ([]string, error) {
+	bufRd := bufio.NewReader(r)
+	chain := make([]string, 0, 8)
+	for {
+		line, err := bufRd.ReadSlice('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+
+		hashStr := string(line[:len(line)-1])
+		if !plumbing.IsHash(hashStr) {
+			return nil, ErrMalformedCommitGraphFile
+		}
+		chain = append(chain, hashStr)
+	}
+	return chain, nil
+}
+
+// OpenChainOrGraph expects a billy.Filesystem representing a .git directory.
+// It will first attempt to read a commit-graph file, before trying to read a
+// commit-graph chain file. If neither are present, an error is returned.
+//
+// An Index and an io.Closer will be returned to close all of the associated files.
+//
+// See: https://git-scm.com/docs/commit-graph
+func OpenChainOrGraph(fs billy.Filesystem) (Index, io.Closer, error) {
+	file, err := fs.Open(path.Join("objects", "info", "commit-graph"))
+	if err == nil {
+		index, err := OpenFileIndex(file)
+		if err == nil {
+			return index, file, nil
+		}
+		_ = file.Close()
+		return nil, nil, err
+	}
+
+	chainFile, err := fs.Open(path.Join("objects", "info", "commit-graph-chain"))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	chain, err := OpenChainFile(chainFile)
+	_ = chainFile.Close()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var index Index
+	fileClosers := make([]io.Closer, 0, len(chain))
```

Comment:
@zeripath I am not sure I understand why we are returning all the file closers. Is there a specific reason why we can't just close them before returning?

Reply:
The files need to remain open for the Index to work. Given we've moved out to a different package I can fix the Index interface to make it be an io.Closer and make this a bit more idiomatic.

Reply:
done